### PR TITLE
Directly insert byte data to BufferBuilder, closes #4722

### LINKED ISSUE #4722: [Suggestion] Improving the BufferBuilder
Currently there is `addVertexData(int[] vertexData)` it would be very helpfull to have a method to insert a byte[] or a bytebuffer directly for performance. But I want to hear more opinions on this matter, espacially because of 1.13 and the render changes there.

### PR DESCRIPTION
**Pull request by mcenderdragon**
This add `addVertexData` but instaed of an int array you can insert a bytebuffer.

This closes #4722

### CODE REVIEW

File: patches/minecraft/net/minecraft/client/renderer/BufferBuilder.java.patch [hunk 1]
```diff
@@ -58,5 +58,13 @@
 +    public boolean isColorDisabled()
 +    {
 +        return this.field_78939_q;
++    }
++
++    public void addVertexData(ByteBuffer buffer)
++    {
++        func_181670_b(buffer.limit());
++        this.field_179001_a.position (this.field_178997_d * this.field_179011_q.func_177338_f());
```
Comment (mezz) [inline patches/minecraft/net/minecraft/client/renderer/BufferBuilder.java.patch]: Shouldn't have a space between `position` and `(`

**Review by mezz — CHANGES_REQUESTED** [review]

**Comment by mezz** [comment]
Is this new method affected by the issue in #4538?
@covers1624 can you give your opinion on this PR?

**Comment by mcenderdragon** [comment]
it is possible, but I can not say for sure, also fixed the space problem ^^.

File: patches/minecraft/net/minecraft/client/renderer/BufferBuilder.java.patch [hunk 2]
```diff
@@ -58,5 +58,13 @@
 +    public boolean isColorDisabled()
 +    {
 +        return this.field_78939_q;
++    }
++
++    public void addVertexData(ByteBuffer buffer)
++    {
++        func_181670_b(buffer.limit());
```
Comment (covers1624) [inline patches/minecraft/net/minecraft/client/renderer/BufferBuilder.java.patch]: This does have the same issue, you need to change this to `func_181670_b(buffer.limit() + this.field_179011_q.func_177338_f());`

**Comment by covers1624** [comment]
I can't really see many use cases for this, but sure. The Unit test needs to be expanded to also test this.

**Comment by mcenderdragon** [comment]
I added the suggested changes. I need this for holograms, I use my own BufferBuilder for the initial rendering and the save the ByteBuffer from there, this way I can insert the data directly and get very high FPS with very large holograms. [Example](https://www.youtube.com/watch?v=Q4lniHfjxWs)

**Comment by covers1624** [comment]
I see, like i said, i personally don't see any use cases, this is definitely one. I would like the UnitTest for BufferBuilder expansion to be expanded also.

File: patches/minecraft/net/minecraft/client/renderer/BufferBuilder.java.patch [hunk 3]
```diff
@@ -58,5 +58,13 @@
 +    public boolean isColorDisabled()
 +    {
 +        return this.field_78939_q;
++    }
++
++    public void addVertexData(ByteBuffer buffer)
```
Comment (bs2609) [inline patches/minecraft/net/minecraft/client/renderer/BufferBuilder.java.patch]: Would suggest `putBulkData`, as that seems to be the "official" name (https://github.com/ModCoderPack/MCPBot-Issues/issues/503).

**Comment by mcenderdragon** [comment]
I think the test works (it has the same output as the other tests)

**Review by covers1624 — APPROVED** [review]

File: patches/minecraft/net/minecraft/client/renderer/BufferBuilder.java.patch (hunk 3 above)
Comment (mcenderdragon) [reply patches/minecraft/net/minecraft/client/renderer/BufferBuilder.java.patch]: Done

**Review by mezz — APPROVED** [review]

**Comment by mezz** [comment]
3 failing commits with "This commit was created on GitHub.com" makes me glad I added the CI to make sure it builds :P 
Please put a little more effort into your PR next time. This code is probably going to be here for a long time.

**Pull request merged**